sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,177 @@
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
+        "Trying to claim key %s before start of the range %s",
+        key,
+        range);
+    lastAttemptedKey = key;
+    // No respective checkArgument for i < range.to() - it's ok to try claiming keys beyond
+    if (!range.getEndKey().isEmpty() && key.compareTo(range.getEndKey()) > -1) {
+      return false;
+    }
+    lastClaimedKey = key;
+    return true;
+  }
+
+  /**
+   * Marks that there are no more keys to be claimed in the range.
+   *
+   * <p>E.g., a {@link DoFn} reading a file and claiming the key of each record in the file might
+   * call this if it hits EOF - even though the last attempted claim was before the end of the
+   * range, there are no more keys to claim.
+   */
+  public synchronized void markDone() {
+    lastAttemptedKey = range.getEndKey();
+  }
+
+  @Override
+  public synchronized void checkDone() throws IllegalStateException {
+    checkState(
+        next(lastAttemptedKey).compareTo(range.getEndKey()) > -1,
+        "Last attempted key was %s in range %s, claiming work in [%s, %s) was not attempted",
+        lastAttemptedKey,
+        range,
+        next(lastAttemptedKey),
+        range.getEndKey());
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("range", range)
+        .add("lastClaimedKey", lastClaimedKey)
+        .add("lastAttemptedKey", lastAttemptedKey)
+        .toString();
+  }
+
+  // Utility methods
+  /**
+   * Calculates the next {@link ByteKey} for a given key by incrementing by one using byte
+   * arithmetic. If the input key is empty it assumes it is a lower bound and returns the 00 byte
+   * array.
+   */
+  @VisibleForTesting
+  static ByteKey next(ByteKey key) {
+    return key.isEmpty()
+        ? ByteKey.of(0)
+        : ByteKey.copyFrom(unsignedCopyAndIncrement(key.getBytes()));
+  }
+
+  // Following methods cloned from org.apache.hadoop.hbase.util.Bytes
+  /**
+   * @return The value of the input incremented by one using byte arithmetic. It treats the input
+   *     byte[] as an unsigned series of bytes, most significant bits first.
+   */
+  private static byte[] unsignedCopyAndIncrement(byte[] input) {
+    byte[] copy = copy(input);
+    if (copy == null) {
+      throw new IllegalArgumentException("cannot increment null array");
+    } else {
+      for (int i = copy.length - 1; i >= 0; --i) {
+        if (copy[i] != -1) {
+          ++copy[i];
+          return copy;
+        }
+        copy[i] = 0;
+      }
+      byte[] out = new byte[copy.length + 1];
+      out[0] = 1;
+      System.arraycopy(copy, 0, out, 1, copy.length);
+      return out;
+    }
+  }
+
+  private static byte[] copy(byte[] bytes) {
```

Comment:
<!--new_thread; commit:d28b9222a4c730ab557091334c555aa8fe89bc67; resolved:0-->
Use Arrays.copyOf() instead of this method?